File: src/main/scala/category/dfs/_700_800/_721_Solution.scala
/**
  * Alipay.com Inc.
  * Copyright (c) 2004-2020 All Rights Reserved.
  */
package category.dfs._700_800

/**
  * 721. Accounts Merge
  */
object _721_Solution {
  def accountsMerge(accounts: List[List[String]]): List[List[String]] = {
    val edge = new scala.collection.mutable.HashSet[(Int, Int)]
    val N = accounts.length
    val emailMapping = new java.util.HashMap[String, Int]
    for (i <- 0 until N) {
      val n = accounts(i).length
      for (j <- 1 until n) {
        if (emailMapping.containsKey(accounts(i)(j))) {
          //          if (emailMapping.get(accounts(i)(j)) != i)
          edge.add((emailMapping.get(accounts(i)(j)), i))
        } else {
          emailMapping.put(accounts(i)(j), i)
        }
      }
    }
    case class Dsu(size: Int) {
      val parent = Array.tabulate(size)(_ => -1)
      
      def find(x: Int): Int = {
        if (parent(x) < 0) return x
        find(parent(x))
      }
      
      def union(i: Int, j: Int): Unit = {
        val left = find(i)
        val right = find(j)
        val s0 = Math.abs(parent(left))
        val s1 = Math.abs(parent(right))
        if (left == right) return
        if (s0 > s1) {
          parent(right) = left
          parent(left) -= s1
        } else {
          parent(left) = right
          parent(right) -= s0
        }
      }
      
      def group(): List[List[Int]] = {
        var ans: List[List[Int]] = Nil
        var i = -1
        parent.map { v =>
          i += 1
          if (v < 0) i -> i else find(v) -> i
        }.groupBy(_._1).foreach { case (_, lst) =>
          ans ::= lst.map(_._2).toList
        }
        ans
      }
    }
    
    /*merge*/
    
    val dsu = Dsu(N)
    edge.foreach { case (src, dst) =>
      dsu.union(src, dst)
    }
    dsu.group().map { lst =>
      val head = lst.head
      val account = accounts(head).head
      account :: lst.flatMap { one =>
        accounts(one).tail
      }.distinct.sorted
    }
  }
}

File: src/main/python/medium/_700_750/_721_Solution.py
# 721. Accounts Merge
from typing import List


class _721_Solution:
    class Dsu:
        def __init__(self, size):
            self.array = [-1] * size

        def find(self, search):
            if self.array[search] < 0: return search
            return self.find(self.array[search])

        def union(self, src, dst):
            ancestor0, ancestor1 = self.find(src), self.find(dst)
            if ancestor0 == ancestor1: return True
            s0, s1 = abs(self.array[ancestor0]), abs(self.array[ancestor1])
            if s0 >= s1:
                self.array[ancestor0] += self.array[ancestor1]
                self.array[ancestor1] = ancestor0
            else:
                self.array[ancestor1] += self.array[ancestor0]
                self.array[ancestor0] = ancestor1

    def accountsMerge(self, accounts: List[List[str]]) -> List[List[str]]:
        if not accounts: return accounts
        lookup, edges = {}, []
        for i, account in enumerate(accounts):
            for j in range(1, len(account)):
                if account[j] in lookup:
                    edges.append([lookup[account[j]], i])
                else:
                    lookup[account[j]] = i
        dsu, ans = self.Dsu(len(accounts)), []
        for edge in edges:
            dsu.union(edge[0], edge[1])
        for i in range(len(accounts)):
            parent = dsu.find(i)
            if parent == i: continue
            accounts[parent] += accounts[i][1:]
        for i in range(len(accounts)):
            parent = dsu.find(i)
            if parent == i:
                emails = list(set(accounts[i][1:]))
                emails.sort()
                ans.append([accounts[i][0]] + emails)
        return ans


if __name__ == '__main__':
    obj = _721_Solution()
    assert obj.accountsMerge(
        [["John", "johnsmith@mail.com", "john_newyork@mail.com"], ["John", "johnsmith@mail.com", "john00@mail.com"],
         ["Mary", "mary@mail.com"], ["John", "johnnybravo@mail.com"]]) == [
               ["John", "john00@mail.com", "john_newyork@mail.com", "johnsmith@mail.com"], ["Mary", "mary@mail.com"],
               ["John", "johnnybravo@mail.com"]]
